File: articles/developmental-psychopathology-risk-resilience-and-adaptation/r/model_developmental_psychopathology.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "developmental_psychopathology_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_developmental_psychopathology.py first.")

panel <- read.csv(data_path)

mixed_model <- lmer(
  adaptation_score ~ time + current_regulation + current_support +
    current_stability + current_risk + cumulative_risk +
    community_support + school_belonging + service_access +
    transition_support + biological_sensitivity +
    current_support:current_stability +
    (1 + time | context_id/child_id),
  data = panel
)

internalizing_model <- lmer(
  internalizing_score ~ time + current_risk + cumulative_risk +
    current_support + current_stability + community_support +
    school_belonging + service_access + biological_sensitivity +
    (1 + time | context_id/child_id),
  data = panel
)

externalizing_model <- lmer(
  externalizing_score ~ time + current_risk + cumulative_risk +
    current_support + current_stability + current_regulation +
    community_support + school_belonging + service_access +
    biological_sensitivity + (1 + time | context_id/child_id),
  data = panel
)

capture.output(
  list(
    adaptation_model = summary(mixed_model),
    internalizing_model = summary(internalizing_model),
    externalizing_model = summary(externalizing_model)
  ),
  file = file.path(outputs_dir, "r_psychopathology_model_summary.txt")
)

trajectory <- panel |>
  group_by(time) |>
  summarize(
    average_adaptation = mean(adaptation_score),
    average_internalizing = mean(internalizing_score),
    average_externalizing = mean(externalizing_score),
    standard_error = sd(adaptation_score) / sqrt(n()),
    lower = average_adaptation - 1.96 * standard_error,
    upper = average_adaptation + 1.96 * standard_error,
    .groups = "drop"
  )

write.csv(trajectory, file.path(outputs_dir, "r_adaptation_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_adaptation_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_adaptation)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper), alpha = 0.15) +
    labs(title = "Synthetic Developmental Psychopathology Trajectories", x = "Time", y = "Average adaptation") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

ggsave(
  filename = file.path(outputs_dir, "r_multifinality_pathways.png"),
  plot = ggplot(trajectory, aes(x = time)) +
    geom_line(aes(y = average_internalizing, linetype = "internalizing"), linewidth = 1) +
    geom_line(aes(y = average_externalizing, linetype = "externalizing"), linewidth = 1) +
    labs(title = "Synthetic Multifinality Pathways", x = "Time", y = "Average pathway score", linetype = "Pathway") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R outputs.")

File: articles/developmental-psychopathology-risk-resilience-and-adaptation/python/model_developmental_psychopathology.py
#!/usr/bin/env python3
"""Model synthetic developmental psychopathology trajectories."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "developmental_psychopathology_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_developmental_psychopathology.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["child_id", "time"]).reset_index(drop=True)
    panel["lag_adaptation"] = panel.groupby("child_id")["adaptation_score"].shift(1)
    panel["lag_internalizing"] = panel.groupby("child_id")["internalizing_score"].shift(1)
    panel["lag_externalizing"] = panel.groupby("child_id")["externalizing_score"].shift(1)
    model_data = panel.dropna(subset=["lag_adaptation"]).copy()

    adaptation_model = smf.ols(
        """
        adaptation_score ~ lag_adaptation + time + current_regulation +
        current_support + current_risk + cumulative_risk +
        current_stability + community_support + school_belonging +
        service_access + transition_support + biological_sensitivity +
        current_support:current_stability
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    internalizing_model = smf.ols(
        """
        internalizing_score ~ lag_internalizing + time + current_risk +
        cumulative_risk + current_support + current_stability +
        community_support + school_belonging + service_access +
        biological_sensitivity
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    externalizing_model = smf.ols(
        """
        externalizing_score ~ lag_externalizing + time + current_risk +
        cumulative_risk + current_support + current_stability +
        community_support + school_belonging + service_access +
        current_regulation + biological_sensitivity
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_psychopathology_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC ADAPTATION MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(adaptation_model.summary().as_text())
        f.write("\n\nINTERNALIZING PATHWAY MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(internalizing_model.summary().as_text())
        f.write("\n\nEXTERNALIZING PATHWAY MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(externalizing_model.summary().as_text())

    trajectory = panel.groupby("time", as_index=False).agg(
        average_adaptation=("adaptation_score", "mean"),
        average_internalizing=("internalizing_score", "mean"),
        average_externalizing=("externalizing_score", "mean"),
        standard_error=("adaptation_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["lower"] = trajectory["average_adaptation"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_adaptation"] + 1.96 * trajectory["standard_error"]
    trajectory.to_csv(OUTPUTS_DIR / "python_adaptation_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(trajectory["time"], trajectory["average_adaptation"], linewidth=2)
    plt.fill_between(trajectory["time"], trajectory["lower"], trajectory["upper"], alpha=0.2)
    plt.xlabel("Time")
    plt.ylabel("Average adaptation score")
    plt.title("Synthetic Developmental Psychopathology Trajectories")
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_adaptation_trajectory.png", dpi=160)
    plt.close()

    plt.figure(figsize=(8, 5))
    plt.plot(trajectory["time"], trajectory["average_internalizing"], marker="o", label="internalizing")
    plt.plot(trajectory["time"], trajectory["average_externalizing"], marker="o", label="externalizing")
    plt.xlabel("Time")
    plt.ylabel("Average pathway score")
    plt.title("Synthetic Multifinality Pathways")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_multifinality_pathways.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("risk_support_profile", as_index=False).agg(
        children=("child_id", "nunique"),
        average_adaptation=("adaptation_score", "mean"),
        average_internalizing=("internalizing_score", "mean"),
        average_externalizing=("externalizing_score", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_risk_support_profiles.csv", index=False)

    print("Wrote Python model outputs.")


if __name__ == "__main__":
    main()
